### PR TITLE
feat(triggers): redact secrets in terminal output live

### LINKED ISSUE #360: Secret redaction: a trigger action that masks keys and tokens in terminal output live
## Motivation

Screen shares, streams, demo GIFs, and pair sessions all leak secrets through the terminal. The trigger engine already watches every output line and can recolour it; a `redact` action that masks matches and reveals them on hover is a one-line config for the user and a real safety feature for anyone who has ever pasted an AWS key into Slack.

## Current state

- `src/triggers.rs`: highlight (painted at render time over visible rows, scrollback included), notify, bell, capture.

## Proposed design

1. New action `redact`: matched spans render as `••••` (keeping the width) at paint time, so the underlying grid is untouched; hover (or the caret in copy mode) shows the real text in a tooltip; copy yields the real value unless `redact.copy = masked`.
2. A built-in default rule set, on by default and listed in Settings: AWS access keys, `sk-`/`ghp_`/`xox`-style tokens, JWTs, private-key PEM blocks, `Authorization: Bearer …`. Users add their own rows in `triggers.json`.
3. Redaction applies to everything downstream of the grid: `croft record`, the browser client, `--solo` guests, and the scrollback-to-editor tab (`Cmd+K D`), which is the point.
4. A status-bar chip counts redactions on screen; click to toggle reveal for 10 s with a confirm.

## Acceptance criteria

- [ ] `echo AKIAIOSFODNN7EXAMPLE` renders masked; hover shows it; copy behaves per setting.
- [ ] A recording made while the rule is on contains the mask, not the key (byte-level test on the cast file).
- [ ] Redaction adds no measurable cost to an idle pane (paint-time only, like highlight).

## Related

- Tracking epic: #383
- Pairs with: #356

### PR DESCRIPTION
**Pull request by vitali87**
## Summary
Closes #360.

- **`redact` trigger action** (`src/triggers.rs`): masks the match — capture group 1 when the regex has one, so `Bearer (\S+)` keeps the header word — as a run of `•` of the same width. Paint-time only: the grid keeps the real text, so an idle pane pays nothing.
- **Built-in rule set**, on by default: AWS access keys, OpenAI-style `sk-`, GitHub `ghp_`/`gho_`/`ghu_`/`ghs_`/`ghr_`/`github_pat_`, Slack `xox*`, JWTs, `Bearer` tokens. Settings row *Terminal: Redact Secrets* (stored inverted as `disable_secret_redaction`) and palette *Terminal: Toggle Secret Redaction* switch it off; the user's own `redact` rows in `triggers.json` are unaffected.
- **Reveal**: a click on a mask pops the real value (same popup as annotations). Palette *Terminal: Reveal Redacted Secrets for 10s* lifts every mask, re-masks on its own, and announces both.
- **Copies** carry the real value unless the rule says `"copy": "masked"`; the `Cmd+K D` scrollback dump is always masked, since that buffer leaves the pane as bytes.
- **Status chip** counts masks on screen (`• 3 redacted`) and turns red while a reveal window is open.
- Docs: KEYBINDINGS.md triggers row, SETTINGS.md key list, the `triggers.json` template. Version 0.1.797 with a matching release note.

### Scope notes
- PEM private-key bodies are not covered: their base64 lines carry no prefix to anchor on, and masking every long base64 line would eat checksums and blobs. The header line is prose, not a secret.
- Everything that leaves the pane as bytes today is masked: the `Cmd+K D` dump, the terminal session transcript written to disk, and the durable command history. `croft record` / the browser client / `--solo` guests (#356, #378) do not exist yet; when they do, they should read through `triggers::mask_text` the same way.
- The chip is not clickable (the status bar has no hit-testing); the reveal is a palette command instead of chip-click-with-confirm.

## Test plan
- [x] `triggers::tests` (13, including: a redact rule never becomes a status-bar hit; the built-in `sk-`/`ghp_` rules skip digit-less words and `Bearer` only masks inside an `Authorization:` header): the `redact` row and its copy mode parse; spans are char-indexed and take group 1; `mask_text` masks per line for the dump and only `copy: masked` rules for the clipboard; the built-in rules mask each key shape and leave prose (`cargo build`, `Bearer of bad news`, a sha256) alone; user rules stack after the built-ins.
- [x] `redacted_spans_paint_as_masks_and_the_real_text_stays_recoverable` (real PTY): the key paints as twenty `•`, text around it is untouched, the chip count is non-zero, a click on the mask recovers the value, a click beside it does not, and a reveal window paints the text as typed.
- [x] `secret_redaction_toggles_the_builtin_rules_and_reveal_expires`: default on, toggle off/on swaps the rule set, the reveal window expires with a redraw, copies respect the per-rule copy mode.
- [x] clippy `-D warnings`, `cargo fmt --check`, doc-ownership check.

## Summary by CodeRabbit

* **New Features**
  * Added automatic terminal masking for common secrets, including API keys, tokens, JWTs, and bearer credentials.
  * Added settings and command palette controls to enable or disable masking and reveal secrets temporarily for 10 seconds.
  * Added click-to-reveal behavior and a status bar count of masked values.
  * Added configurable redaction triggers with optional masked clipboard copying.
* **Documentation**
  * Updated terminal keybinding, settings, and release documentation.
* **Chores**
  * Bumped the package version to 0.1.797.

### CODE REVIEW

**Comment by vitali87** [comment]
## 🤖 Fallback review — Claude Opus 5

CodeRabbit rate-limited on this PR ("Review limit reached"), so this local review of equivalent shape stands in. It reviewed `accebfd`; the follow-up commit addresses every finding below.

**Walkthrough**

Adds a `redact` trigger action that masks matched spans as `•` runs at paint time, leaving the grid text intact, plus a built-in rule set (AWS, `sk-`, `ghp_`, `xox`, JWT, Bearer) enabled by default behind an inverted `disable_secret_redaction` pref. Masks can be lifted per-span by clicking (hover popup) or globally for 10s via a palette command, with a status-bar chip counting on-screen masks. The `Cmd+K D` scrollback dump always masks; clipboard copies carry the real value unless a rule opts in with `"copy": "masked"`.

| File | Change |
| --- | --- |
| `src/triggers.rs` | `TriggerAction::Redact`, `MASK`, `copy_masked`, `redact_spans`, `mask_text`, `builtin_redactions`, `with_builtin_redactions`, `has_redactions`; `has_highlights`/`has_events` reclassified |
| `src/widgets/terminal.rs` | `TrigCell`; mask painting; `reveal_redactions`, `redacted_on_screen`, `redacted_at` |
| `src/app/mod.rs` | `secret_redaction`, `redaction_reveal_until`, toggle/reveal/tick, `load_trigger_set`, status chip, copy + dump masking, click-to-reveal |
| `src/prefs.rs`, `src/widgets/command_palette.rs`, tests, docs, `Cargo.toml`, `src/release_notes.rs` | pref + saver, two commands, coverage, docs, 0.1.797, replaced note |

**Findings**

- **`src/triggers.rs:566-570`** **[blocker/security]** The reader-thread scanner emits a `TriggerHit` for every `Redact` rule, carrying the raw secret, and the app prints it in the status bar. `complete_line` skips only `Highlight`; redact rules have `message: None`, so `interpolate` returns the whole match, and the app's `_ =>` arm sets `self.status = "Trigger in …: <secret>"`. Verified by a compiled probe: `TriggerHit { action: Redact, message: "AKIAIOSFODNN7EXAMPLE", … }`. The scan is not gated on `has_events()`. Net effect: `echo AKIA…` masks the pane line and prints the key verbatim in the status bar, on by default. Fix: skip `Redact` in the scanner; add a regression test asserting zero hits.
- **`src/app/mod.rs:12075-12082`** **[major/security]** The terminal session store persists up to 200 unmasked scrollback lines per pane to disk (`grid_lines()` → `terminal_session::save_for_root`, plain JSON), replayed next launch. Route the transcript through `mask_text(…, false)` like the dump.
- **`src/app/mod.rs:26246-26262`** **[major/security]** The durable command history persists the typed command line unmasked — exactly where secrets appear (`export AWS_ACCESS_KEY_ID=…`, `curl -H "Authorization: Bearer …"`), on disk and cross-session. Mask it or document the exclusion.
- **`src/triggers.rs:253`** **[major/correctness]** The `Bearer` rule masks ordinary English words: `"Bearer responsibilities matter"` → `"Bearer •••••••••••••••• matter"`; `sk-something-like-this-long` is masked too. On by default. Anchor on `Authorization:`; require credential-shaped material for `sk-`.
- **`src/widgets/terminal.rs:4259`** **[minor/correctness]** `redacted_on_screen` goes stale when the pane does not paint (panel hidden, another tab, zero-width column); the chip then claims masks with nothing on screen. Reset it wherever `last_area` is reset.
- **`src/widgets/terminal.rs:2387-2389`** **[minor/correctness]** `redacted_at` uses exact column equality, so a click on a wide char's spacer column inside a masked span returns `None`; `line_text_at` resolves the same case with `rposition(|&gc| gc <= c)`.
- **`src/app/mod.rs:33385-33400`** **[minor/correctness]** The redaction popup precedes click-to-move on the prompt row; defensible, but worth a comment naming the trade-off.

**Test coverage**

- The scanner/redact leak (fails on the branch today).
- Session-store masking (byte-level, as #360 asked).
- Command-history masking.
- Bearer/`sk-` false positives in the prose list.
- Chip staleness after hiding the panel.
- Wide-char click.

**Nitpicks**

- `redact_spans` computes a `chars().count()` prefix per match per row per frame; consistent with `highlight_spans`, but now on by default.
- `mask_text` double-walks the trigger list.
- `terminal_text_for_copy` takes `String` by value.
- The PEM scope note understates what a stateful header-anchored rule could do.

**Scores**

**CodeRabbit score (diff quality) — 1/5** — one blocker: the built-in rules print the secrets they mask into the status bar.

**Greptile score (codebase fit) — 2/5** — the two persistent on-disk sinks (session transcript, command history) were not routed through `mask_text`, plus a default-on regex that corrupts prose.

VERDICT: REQUEST_CHANGES

**Comment by vitali87** [comment]
## 🤖 Fallback re-review (round 2) — Claude Opus 5

CodeRabbit is still rate-limited on this PR, so this local review of equivalent shape stands in. It reviews `be11601`; the follow-up commit after this one addresses everything below.

## 1. Walkthrough

The follow-up addresses all seven earlier findings: redact rules are excluded from the reader-thread scanner, the session transcript and command history are routed through `mask_text`, the `Bearer` rule is anchored on `Authorization:`, three prefix rules gain a `require_digit` gate, `redacted_on_screen` is zeroed at every `last_area` reset, and `redacted_at` resolves wide-char spacer columns. The blocker and both persistence majors are genuinely fixed and verified at the source. But the sweep for sinks was incomplete — a fourth clipboard path copies raw command output with no masking — and the `require_digit` gate is an over-correction that silently drops real tokens into the very sinks just fixed.

| File | Change |
| --- | --- |
| `src/triggers.rs` | `require_digit`; `Bearer` re-anchored; `complete_line` skips `Redact`; 2 tests |
| `src/app/mod.rs` | transcript and history via `mask_text`; 6× `redacted_on_screen = 0`; ordering comment |
| `src/widgets/terminal.rs` | `redacted_at` uses `rposition` for the char index |
| `src/app/tests.rs` | chip-staleness assertion |

### Earlier findings — all seven resolved (verified at source)

Scanner skips `Highlight | Redact` with a test asserting zero hits; transcript lines and history `cmd` go through `mask_text(…, false)`; `Bearer` anchored on `(?i)\bauthorization: *bearer`; all six `last_area` resets paired with the chip reset; spacer columns resolve via `rposition`; the ordering trade-off is documented. Sinks verified clean: CAPTURES (in-memory, unreachable for built-ins), the hover popup (render-only), diagnostics scans, the re-exec handoff (excludes terminal panes), the FinishedCommand notification (label/exit/duration only). Transcript masking cost is acceptable (structural-change/quit call sites, ≤200 lines per pane, early-out when no redact rule).

## 2. Findings

- **`src/app/mod.rs:26435`** **[major/security]** `copy_command_output` copies a command's entire output to the clipboard with no masking — the fourth clipboard writer of terminal text and the only one that skips `terminal_text_for_copy` (the three selection copies all use it). Reachable from the decoration menu's Copy Output and `⌘K ⇧C`. Fix: route through `terminal_text_for_copy`.
- **`src/triggers.rs:256,259,263`** **[major/security]** The `require_digit` gate drops real tokens: `ghp_…ABCDEFGHIJ`, `xoxb-abcdef…`, `sk-proj-abc…` no longer match (false-negative rate up to ~1 in 6 for `xox`), and they now flow unmasked into the transcript and history this commit fixed. `gh[pousr]_`/`github_pat_` and `xox` never needed the gate (their prefixes anchor, no hyphens); the branch's own test enshrines the miss. Fix: drop the gate; constrain `sk-` by shape (at most one labelled segment) instead.
- **`src/app/mod.rs:30307`** **[minor/security]** Quick-select copies (and on shift pastes) a matched token unmasked; the hex/number matchers can land inside a masked span, so a fragment of a secret is one keystroke from the clipboard. Fix: route `hit.text` through `terminal_text_for_copy`.
- **`src/triggers.rs:271`** **[minor/correctness]** `(\S{16,})` is greedy to the end of the run, so `curl -H 'Authorization: Bearer …'` masks the closing quote too. Fix: `([^\s'"]{16,})`.
- **`src/widgets/terminal.rs:2389`** **[minor/correctness]** `rposition(|&gc| gc <= vc)` has no upper bound: a click in the blank area right of the last character resolves to the last char, so clicking empty space after a masked token pops the secret. Fix: reject when `vc` exceeds the last mapped column plus its width.

## 3. Test coverage

- Session-transcript and command-history masking ship with no test; #360 asked for a byte-level assertion on the on-disk store.
- `copy_command_output`: nothing asserts what reaches the clipboard.
- The wide-char spacer click: the test uses an ASCII key.
- `require_digit` false negatives; the bearer anchor against quoted `curl -H` invocations.

The chip-staleness test is well built (fails if the reset is removed).

## 4. Nitpicks

- `let triggers = self.triggers.clone();` deserves a comment on why the Arc clone.
- The `(pattern, digit)` tuple table reads worse than a named struct.
- The doc comment still says the prefix is the only gate.
- `builtin_redactions()` recompiles seven regexes per call; a `OnceLock` would do.

## 5. Scores

Version 0.1.797 vs main 0.1.795 (skips .796, which #389 holds); release notes correctly replaced.

**CodeRabbit score (diff quality) — 2/5** — the blocker and both persistence majors are properly fixed, but two majors remain: an unmasked clipboard path the sweep missed, and a digit gate that re-opens the leak beside the one it closed.

**Greptile score (codebase fit) — 2/5** — `copy_command_output` diverges from the three sibling clipboard writers; the two headline persistence fixes ship with no test.

VERDICT: REQUEST_CHANGES

**Comment by vitali87** [comment]
## 🤖 Fallback re-review (round 3) — Claude Opus 5

CodeRabbit is still rate-limited on this PR, so this local review of equivalent shape stands in. It reviews `c65a63a`; the follow-up commit after this one addresses everything below.

## 1. Walkthrough

This commit routes the two missed clipboard paths through `terminal_text_for_copy`, replaces the `require_digit` gate with a shape constraint on `sk-`, excludes quotes from the bearer capture, adds a bound to `redacted_at`, and adds a byte-level session-store test. Three of the five round-2 findings are genuinely fixed; the `redacted_at` bound is dead code that cannot fire, and the `sk-` shape rule trades one false-negative class for a larger one — it no longer matches the Anthropic key format named in the module's own doc comment.

| File | Change |
| --- | --- |
| `src/app/mod.rs` | `copy_command_output` and the quick-select arm wrapped in `terminal_text_for_copy` |
| `src/triggers.rs` | `require_digit` removed; `sk-` rewritten as a shape rule; bearer capture `([^\s'"]{16,})`; tests |
| `src/widgets/terminal.rs` | `redacted_at` gains a `vc > last + 1` early return |
| `src/app/tests.rs` | `the_session_store_holds_the_mask_not_the_key` |

### Round-2 findings — status

| Finding | Status |
| --- | --- |
| `copy_command_output` unmasked | Structurally fixed; functionally inert for built-ins because they are `copy_masked: false` by design (documented). |
| `require_digit` dropping real tokens | Resolved for `gh`/`xox`; regressed for `sk-` (below). |
| Quick-select copy/paste unmasked | Structurally fixed; inert twice over — the hint is a fragment no anchored rule can re-match (below). |
| Bearer capture swallowing a quote | Resolved, with a test. |
| `redacted_at` blank cells past the last char | **Not resolved** — the guard is dead code (below). |
| Missing byte-level tests | Half: the session store has one; command history still none. |

## 2. Findings

- **`src/widgets/terminal.rs:2389-2393`** **[major/correctness]** `row_text_and_cols` pushes a char for every non-spacer column, blanks included, so `colmap.last()` is `ncols-1` and `vc` is bounded by the grid width: `vc > last + 1` can never hold. Clicking blank space right of a masked token still pops the secret. Bound on the last non-blank char instead, with a test that clicks past the row's text.
- **`src/triggers.rs:249`** **[major/correctness]** `\bsk-(?:[A-Za-z0-9_]+-)?…` permits at most one label; `sk-ant-api03-…` (Anthropic, named in the crate doc) and `sk-or-v1-…` (OpenRouter) do not match and flow unmasked into the sinks the previous commit masked. Allow repeated short labels.
- **`src/triggers.rs:249`** **[minor/correctness]** The same rule masks `sk-preprocessingtransformerspipeline`; the prose list has no long-final-segment case.
- **`src/app/mod.rs:30309`** **[minor/security]** Quick-select hints are fragments (`\b[0-9a-f]{7,40}\b`, `\b[0-9]{4,}\b`) that can sit inside a secret; re-matching the fragment cannot mask it. Drop hints that overlap a redacted span on the source row.
- **`src/triggers.rs:271`** **[minor/correctness]** The bearer rule misses the JSON form (`"Authorization": "Bearer …"`) and still captures a trailing comma.
- **`src/triggers.rs:239-244`** **[nit/docs]** Over-long reflowed doc line.

## 3. Test coverage

The `redacted_at` bound has no test (which is why the dead guard shipped); `copy_command_output` and quick-select assert nothing about the clipboard; `sk-ant-api03-…`/`sk-or-v1-…` absent from the assertions; command-history masking untested; the bearer JSON form untested. The new session-store test is a real positive control with sound cleanup; its 10 s poll loop is unnecessary.

## 4. Nitpicks

The commit subject overstates ("mask every clipboard path"); `builtin_redactions()` still recompiles seven regexes per call — `LazyLock` is the in-repo precedent.

## 5. Scores

Version 0.1.797 against main's 0.1.795 (skipping .796, held by #389); release notes replaced, not appended.

**CodeRabbit score (diff quality) — 2/5** — a guard that cannot execute, and a shape rule that stops matching the Anthropic key format the crate doc advertises.

**Greptile score (codebase fit) — 2/5** — the fix pattern matches the sibling clipboard writers, but three of five changes ship untested, and the two testable ones are the ones that do not work.

VERDICT: REQUEST_CHANGES

**Comment by vitali87** [comment]
## 🤖 Fallback re-review (round 4) — Claude Opus 5

CodeRabbit is still rate-limited on this PR, so this local review of equivalent shape stands in. It reviews `d490bff`, rebased onto main (which now carries #389). The follow-up commit after this one addresses everything below.

## 1. Walkthrough

This commit rewrites the `redacted_at` bound against the row's trimmed text, widens the `sk-` shape to three labels behind a new `needs_entropy` gate, teaches the bearer rule the JSON form and trailing punctuation, drops quick-select hints that overlap a masked span, and moves the built-in rules behind a `LazyLock`. Four of the six round-3 findings are genuinely fixed and verified by compiled probe. The `redacted_at` bound is still not demonstrably live — the new test passes identically with the guard deleted, because its click column lands outside the pane and is rejected by `cell_at` before the guard is reached.

| File | Change |
| --- | --- |
| `src/triggers.rs` | `needs_entropy` gate; `sk-` widened; bearer JSON form; `LazyLock`; tests |
| `src/widgets/terminal.rs` | `redacted_at` bound on the trimmed text; one assertion |
| `src/app/mod.rs` | quick-select retains only hints disjoint from redacted spans |
| `src/app/tests.rs` | poll loop dropped from the session-store test |

### Round-3 findings — status

Dead bound: **not resolved** (see below). `sk-` multi-label: **resolved**, probe-verified. Lowercase hyphenated prose: **resolved**. Quick-select fragments: **resolved for single-row matches**; wrapped tails not checked. Bearer JSON/comma: **resolved**, six forms probed. `LazyLock`: **resolved**.

## 2. Findings

- **`src/widgets/terminal.rs:2392-2394`** **[major/correctness]** The branch's own test passes with the guard deleted: the row renders 60 wide, `past_text` is 63, `last_inner.width` is 58, so `cell_at(63, y)` returns `None` before the guard runs. Fix: click inside the pane past a masked token that is the row's last text.
- **`src/triggers.rs:263,328-333`** **[minor/security]** `needs_entropy` checks the whole match including the label, so `api03` vouches for a lowercase-only body while `sk-abcdef…` (all-lowercase, no label) is dropped. Apply the test to the body after the last label hyphen.
- **`src/app/mod.rs:30269-30277`** **[minor/security]** The retain checks only the head span; a match that wraps the pane edge carries `tail` segments that can sit inside a masked span. Check the tail too.
- **`src/app/mod.rs:30271`** **[nit/perf]** `redact_spans` re-runs per surviving match on the same row.

Verified: `m.row` indexes the same `lines` vector `visible_lines_and_clock` returns; `m.start`/`m.len` are per-row char indices, the same space as `RedactSpan`.

## 3. Test coverage

The `redacted_at` bound (the headline fix) is untested in effect; `needs_entropy` false negatives on an all-lowercase body; the quick-select retain ships with no test; wrapped tails; command-history masking (unchanged since round 2). The session-store test remains a real positive control.

## 4. Nitpicks

The commit message overstates the bound test; `RULES.clone()` is cheap (`Regex` is `Arc`-shared) but the doc undersells the per-`with_builtin_redactions` clone; the `(pattern, bool)` tuple table; the `needs_entropy` doc's "never has either" is saved by the label cap, not the gate.

## 5. Scores

Version 0.1.797 vs main 0.1.796 — correct; `src/release_notes.rs` replaced, not appended.

**CodeRabbit score (diff quality) — 3/5** — two regex majors properly fixed and probe-verified, but the `redacted_at` bound ships for the second round with a test that cannot detect its removal.

**Greptile score (codebase fit) — 3/5** — the regex work and the `LazyLock` match in-repo precedent and the quick-select coordinate spaces align; the wrapped-match tail is unhandled and the central fix is unverified by its own test.

VERDICT: REQUEST_CHANGES

**Comment by vitali87** [comment]
## 🤖 Fallback re-review (round 5) — Claude Opus 5

CodeRabbit is still rate-limited on this PR, so this local review of equivalent shape stands in. It reviews `3d52eb8`. The follow-up commit after this one addresses everything below.

## 1. Walkthrough

This commit moves the mask-click probe inside the pane, applies the entropy gate to the key body rather than the whole match, and extends the quick-select retain over wrapped `tail` segments. Two of the four round-4 findings are genuinely resolved; the headline `redacted_at` bound is still not proven by its own test — built in a throwaway worktree, the test passes byte-identically with the bound deleted, for the third round running. The new entropy gate also writes a real false negative into the test suite as expected behaviour.

| File | Change |
| --- | --- |
| `src/widgets/terminal.rs` | bound test rewritten: second `tail` output row, probe at `tx + 24` inside the pane |
| `src/triggers.rs` | entropy gate reads the segment after the last hyphen; 2 tests |
| `src/app/mod.rs` | retain extracted to a `clear` closure over head + every `tail` segment |
| `src/app/tests.rs` | `quick_select_skips_hints_inside_masked_spans` |

### Round-4 findings — status

Bound test outside the pane: **not resolved** — the click now resolves (`cell_at=Some((3, 29))`), but `row_text_and_cols` maps every non-spacer column, blanks included, so `colmap` is dense and `ci == vc == 29`, a blank past the mask (chars 5..24): no span covers it with or without the bound. Entropy on the whole match: **resolved structurally**, but `sk-ant-api03-<all-lowercase>` is now left unmasked and the new test asserts that miss is correct while the crate doc names the format as supported. Wrapped `tail`: **resolved**, coordinate-correct. Quick-select test: **vacuous** — its key line `key AKIA1234567890123456 end` matches none of the 13 quick-select patterns (no word boundary before the digit run), so the retain filters nothing. Command-history masking: still untested.

## 2. Findings

- **`src/widgets/terminal.rs:2392-2394`** **[major/correctness]** The bound cannot fire: the only case where `rposition` steps back is a wide char's spacer column, which resolves to the wide char itself. Either find a reachable case or document it as unreachable rather than sell it as a fix.
- **`src/triggers.rs:329,772-776`** **[major/security]** Body-only entropy leaves `sk-ant-api03-<lowercase>` unmasked into the transcript and history this PR protects. Keep the label cap as the prose defence and drop the gate for labelled matches.
- **`src/app/tests.rs:20949-20977`** **[major/testing]** The quick-select test passes with the entire retain deleted. Use a masked token that is itself a hint-matchable run at a word boundary and confirm by deletion.
- **`src/app/mod.rs:30269-30281`** **[minor/testing]** The `tail` branch is exercised by nothing (both transcript lines are short and unwrapped).
- **`src/app/mod.rs:30271`** **[nit/perf]** `redact_spans` re-runs per surviving match and now per tail segment.

## 3. Test coverage

The bound in effect; the retain in effect; wrapped tails; the entropy false negative is tested as correct; command-history masking (since round 2).

## 4. Nitpicks

The commit subject overstates "prove"; the `needs_entropy` doc credits the gate for what the label cap does; the probe guard checks the buffer, not the pane; the `(pattern, bool)` table.

## 5. Scores

Version 0.1.797 vs main 0.1.796 — correct; `src/release_notes.rs` replaced, not appended.

**CodeRabbit score (diff quality) — 2/5** — three majors, all test-quality: the bound ships a third round with a test that cannot detect its removal, the quick-select test is vacuous, and the entropy fix enshrines a real false negative.

**Greptile score (codebase fit) — 3/5** — the `tail` retain matches `find_matches_wrapped`'s contract and the integration is consistent; the verification story does not hold up.

VERDICT: REQUEST_CHANGES

**Comment by vitali87** [comment]
## 🤖 Fallback re-review (round 6) — Claude Opus 5

CodeRabbit is still rate-limited on this PR, so this local review of equivalent shape stands in. It reviews `7925b60`.

## 1. Walkthrough

This commit deletes the `redacted_at` bound that three rounds of tests could not kill, replacing it with a comment explaining why it was unreachable; narrows the entropy gate to unlabelled `sk-` matches so a labelled key masks whatever its body looks like; and rebuilds the quick-select test around a token that is itself a hint-matchable hex run. All three round-5 majors are genuinely resolved, and the quick-select test is now a real positive control — verified by deletion in a throwaway worktree. No new defects found.

| File | Change |
| --- | --- |
| `src/widgets/terminal.rs` | `redacted_at` bound removed; comment states why a blank click cannot step back into a mask |
| `src/triggers.rs` | `labelled` derivation gates the entropy test; doc rewritten; `sk-learn-…` moved to the masked assertions |
| `src/app/tests.rs` | quick-select test masks `token deadbeefcafe0123`, asserts the hex run is absent from hints |

### Round-5 findings — status

| Finding | Status | Evidence |
| --- | --- | --- |
| bound unreachable, sold as a fix | **Resolved** | Deleted; claim verified at source (`row_text_and_cols` pushes a char for every non-spacer cell, so the colmap is dense; the only backward step is a spacer resolving to its own wide char). |
| body-only entropy leaves `sk-ant-api03-<lowercase>` unmasked | **Resolved** | `labelled` short-circuits the gate; asserted. |
| quick-select test vacuous | **Resolved, verified by deletion** | With the `matches.retain` block removed the test FAILS: `["deadbeefcafe0123", "cafebabe0123456"]`. |
| wrapped `tail` untested | Not resolved. |
| command-history masking untested | Not resolved (since round 2). |

## 2. Findings

None at blocker, major, or minor severity. Checked and clean: `labelled` is computed but never consulted for user rules (`needs_entropy` is false for them); removing the bound loses nothing for a spacer click (matches `line_text_at`); the retained prose case still passes (13/13).

## 3. Test coverage

Wrapped `tail` segments; command-history masking; the unlabelled-lowercase trade is documented but unbounded by a test; a wide-char spacer click inside a mask (verified by simulation, not in the suite).

## 4. Nitpicks

`labelled`/`body` computed for every rule; `matches('-').count() > 1` is correct only by the regex's construction; the `(pattern, bool)` table; `redact_spans` re-runs per match/segment.

## 5. Scores

Version 0.1.797 vs main 0.1.796 — correct; `src/release_notes.rs` replaced, not appended.

**CodeRabbit score (diff quality) — 5/5** — all three round-5 majors resolved; the headline test now fails by deletion for the right reason, and the removed bound is justified at the source.

**Greptile score (codebase fit) — 4/5** — the dense-colmap reasoning aligns `redacted_at` with `line_text_at`; two carried gaps remain untested.

VERDICT: APPROVE

File: src/prefs.rs [hunk 1]
```diff
@@ -380,6 +386,13 @@ pub fn save_format_on_save(enabled: bool) -> Result<()> {
     prefs.save(&path)
 }
 
+pub fn save_disable_secret_redaction(disabled: bool) -> Result<()> {
+    let path = config_path();
+    let mut prefs = Prefs::load(&path).unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (vitali87) [reply src/prefs.rs]: Fixed in 55d1f8e: the saver now goes through `prefs_for_update` (split into a path-taking `set_disable_secret_redaction`), with a test that a malformed config is refused byte-for-byte and fails against the old `unwrap_or_default`.
